[cas] Add a clang-cas-test utility with option to dump compile-job cache keys

benlangmuir opened the pull request:
Usage:

```
clang-cas-test -print-compile-job-cache-key [-cas PATH] <KEY>
```

This makes it easier to debug what went into a given compile-job cache
key and to compare it to another key by printing out the command-line,
filesystem, and other entries in a format that is amenable to diff.

Incidentally, factor the code for creating a cache key into the same
file as the printer.

benlangmuir committed:
[cas] Add a clang-cas-test utility with option to dump compile-job cache keys

Usage:
    clang-cas-test -print-compile-job-cache-key [-cas PATH] <KEY>

This makes it easier to debug what went into a given compile-job cache
key and to compare it to another key by printing out the command-line,
filesystem, and other entries in a format that is amenable to diff.

Incidentally, factor the code for creating a cache key into the same
file as the printer.

benlangmuir commented:
@swift-ci please test

dexonsmith commented on clang/test/CAS/print-cc1-cache-key.c:
```diff
@@ -0,0 +1,34 @@
+// RUN: not c-index-test core -print-cc1-cache-key -cas %t/cas -key @%t/casid 2>&1 | FileCheck %s -check-prefix=NOT_A_KEY
+// NOT_A_KEY: not a valid cache key
+//
+// RUN: c-index-test core -print-cc1-cache-key -cas %t/cas -key @%t/cache-key | FileCheck %s
```
I wonder if this really belongs in `llvm-cas` instead, something like:
```
// RUN: llvm-cas -print-node llvm
```
That would make sense if:
- The cache key were built using `NodeBuilder` (does not exist), using `.addField(StringRef, StringRef)` and `.addField(StringRef, ObjectRef)`
- We had a `NodeReader` that could do a pretty print.

I'm not askig you take all that on, but it does seem wrong that we'd have this logic in c-index-test when every cache key will want (basically) the same treatment.

benlangmuir replied:
I agree that having a way to print nodes is good, but I'm not sure it replaces a clang-specific printer. Some considerations:

* I don't want to print file contents by default, but I do want to print the other string contents by default. They look the same in the current tree. What would a generic node printer do?
* Command-line arguments
  *  They're null-separated, and I need to print them separately. Unless NodeReader knows about that pattern, it will not be a useful printer for these.
  * I'm breaking lines in the output on arguments that start with "-" which I find gives a better presentation for a command-line (e.g. "-x c" instead of "-x\n-c", but I would be willing to give that up for just each argument on its own line if it came down to it. I'm also adding trailing `\` which lets you copy and paste the command-line easily.

If you think we can do all that in a generic way without llvm-cas needing to know what clang is doing, then I'm open to switching to that when it's available.  If not, then I think this is still useful to be clang-specific.

akyrtzi replied:
I agree with Ben that as we add clang-specific CAS objects it's useful to have domain-specific printers.
Another way I would use this later on is for having sensible printing for the "include-tree" CAS object.

The thing I'm not clear about is whether we should add such stuff on top of `c-index-test` or whether we should introduce a `clang-cas-test` testing utility, WDYT?

dexonsmith commented on clang/test/CAS/print-cc1-cache-key.c:
```diff
@@ -0,0 +1,34 @@
+// RUN:   -e "s/^.*miss for '//" \
+// RUN:   -e "s/' .*$//" > %t/cache-key
+//
+// RUN: not c-index-test core -print-cc1-cache-key -cas %t/cas 2>&1 | FileCheck %s -check-prefix=NO_KEY
```
Should this be `-print-compile-job-cache-key`?

benlangmuir commented:
@swift-ci please test

cachemeifyoucan reviewed (commented):
I think it makes sense to have a clang specific tools to print the cache keys. Moving Tree visitor is fine but we are going to redo it after Duncan's rewrite.

But we should put the logics for computing cache key and print cache key together, instead of leaving the printing logic in c-index-test main.

benlangmuir commented:
Updated:
* Created `clang-cas-test` utility instead of piggy-backing c-index-test
* Refactored the code for creating cache keys to live in the same file as the printing code. This also deduplicated it between cc1_main and cc1depscan_main.

benlangmuir commented:
@swift-ci please test

benlangmuir committed:
[cas] Ensure cache key command-line contains -cc1 argument consistently

Fixes discrepancy between cc1_main, which was not adding -cc1 to the
cache key's command-line, and cc1depscan_main, which was. It's also more
useful to include the cc1 for debugging.

benlangmuir commented:
@swift-ci please test
